### PR TITLE
Create the Users class

### PR DESCRIPTION
**Pull request by adrianostanley**
# Summary

This PR adds the `Users` class.

## QA

- [ ] Code review

## Before merge

- [ ] I have confirmed these changes in each supported minor WooCommerce version

### CODE REVIEW

File: src/Admin/Users.php [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	/** @var string customer role to match with role parameter */
+	private $customer_role = 'customer';
+
+
+	/**
+	 * Renders a Notice object if the users customer role message is enabled.
+	 *
+	 * @since 1.1.0-dev.1
+	 */
+	private function add_admin_notices() {
```
Comment (unfulvio) [inline src/Admin/Users.php:50]:
```suggestion
	public function add_admin_notices() {
```

callbacks to WordPress hooks cannot be private, only public

when adding a callback that is meant for internal use (most of the times with hooks), you can add an `@internal` tag to phpdoc to signal that

Comment (adrianostanley) [reply src/Admin/Users.php]: > callbacks to WordPress hooks cannot be private, only public
> 
> when adding a callback that is meant for internal use (most of the times with hooks), you can add an `@internal` tag to phpdoc to signal that

Yeah, my bad. Lack of attention. Thanks for pointing that!

File: src/Admin/Users.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+		] );
+		$notice->set_title( __( 'Show your customers you care by keeping in touch!', 'sv-wc-jilt-promotions' ) );
+		$notice->set_content( __( 'Use Jilt to send welcome emails, thank customers for purchases, and encourage lapsed customers to shop again. Do you want to install Jilt for WooCommerce to start emailing customers? You’ll be able to connect with one click!', 'sv-wc-jilt-promotions' ) );
+
+		$notice->render();
+	}
+
+
+	/**
+	 * {@inheritDoc}
```
[the remark this answers is not in the document]
Comment (adrianostanley) [reply src/Admin/Users.php]: Yeah, makes sense. It's fixed now.

File: src/Admin/Users.php [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+		// determines if the ID of the current screen is the Users screen ID
+		$is_user_current_screen = isset( $current_screen->id ) && $current_screen->id === $this->users_screen_id;
+
+		// determines if the role parameter is set to customer
+		$is_role_parameter_customer = isset( $_GET['role'] ) && $_GET['role'] === $this->customer_role;
+
+		if ( $is_user_current_screen && $is_role_parameter_customer ) {
+			Messages::enable_message( $this->customer_role_message_id );
+		}
+	}
```
Comment (N-Molham) [inline src/Admin/Users.php:120]: When do this callback get called? I know from the Doc it's callback for admin_enqueue_scripts action but I don't see where it adds the action itself => add_action( 'admin_enqueue_scripts'. ... )

File: src/Admin/Users.php [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+	/**
+	 * Adds the necessary action & filter hooks.
+	 *
+	 * @since 1.1.0-dev.1
+	 */
+	protected function add_prompt_hooks() {
+
+		if ( ! Messages::is_message_dismissed( $this->customer_role_message_id ) ) {
+			add_action( 'admin_notices', [ $this, 'add_admin_notices' ], 15 );
+		}
```
Comment (unfulvio) [inline src/Admin/Users.php:79]:
```suggestion
		if ( ! Messages::is_message_dismissed( $this->customer_role_message_id ) ) {
			
			add_action( 'admin_notices', [ $this, 'add_admin_notices' ], 15 );
			
			add_action( 'admin_enqueue_scripts', [ $this, 'maybe_enable_users_customer_role_message' ] );
		}
```

perhaps this is missing?

File: src/Admin/Users.php (hunk 3 above)
Comment (unfulvio) [reply src/Admin/Users.php:120]: maybe here? https://github.com/skyverge/wc-jilt-promotions/pull/8/files#r469955943

File: src/Admin/Users.php (hunk 4 above)
Comment (N-Molham) [reply src/Admin/Users.php:79]: Yeah, I think that is it

File: src/Admin/Users.php [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+		$notice = new Notice();
+		$notice->set_message_id( $this->customer_role_message_id );
+		$notice->set_actions( [
+			'label'   => __( 'Email my customers', 'sv-wc-jilt-promotions' ),
+			'name'    => 'email-my-customers-cta',
+			'url'     => 'https://www.skyverge.com/go/email-customers',
+			'primary' => true,
+			'type'    => 'link,'
+		] );
```
Comment (wvega) [inline src/Admin/Users.php:62]: @adrianostanley we need to define an array of options here. The `set_actions()` method can handle an array of arrays now: https://github.com/skyverge/wc-jilt-promotions/commit/82b16e2d737973e3128fabcf52146bb514b32ca1.

```suggestion
		$notice->set_actions( [
			[
				'name'    => 'email-my-customers-learn-more',
				'label'   => __( 'Learn more', 'sv-wc-jilt-promotions' ),
				'url'     => 'https://www.skyverge.com/go/email-customers',
				'type'    => Notice::ACTION_TYPE_LINK
			],
			[
				'name'    => 'email-my-customers-cta',
				'label'   => __( 'Email my customers', 'sv-wc-jilt-promotions' ),
				'primary' => true,
				'type'    => Notice::ACTION_TYPE_BUTTON
			]
	   ] );
```

Comment (adrianostanley) [reply src/Admin/Users.php:62]: Sure thing.

File: src/Admin/Users.php [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+				'name'    => 'email-my-customers-cta',
+				'label'   => __( 'Email my customers', 'sv-wc-jilt-promotions' ),
+				'primary' => true,
+				'type'    => Notice::ACTION_TYPE_BUTTON
+			]
+		] );
+		$notice->set_title( __( 'Show your customers you care by keeping in touch!', 'sv-wc-jilt-promotions' ) );
+		$notice->set_content( __( 'Use Jilt to send welcome emails, thank customers for purchases, and encourage lapsed customers to shop again. Do you want to install Jilt for WooCommerce to start emailing customers? You’ll be able to connect with one click!', 'sv-wc-jilt-promotions' ) );
+
+		$notice->render();
```
Comment (wvega) [inline src/Admin/Users.php:73]: @adrianostanley I was giving this one another try after merging the latest from the release branch and I realized we will need to wrap this part in an if to check that the message is enabled before rendering the notice.

Comment (adrianostanley) [reply src/Admin/Users.php]: Done.

**Comment by adrianostanley** [comment]
> Looks good @adrianostanley. Thank you for the latest updates! LGTM.
> 
> I added a small tweak for the notice's message ([a9d2a3c](https://github.com/skyverge/wc-jilt-promotions/commit/a9d2a3c9db698a0b23e8a83e36c67be89d009823)) based on this [conversation on the pitch](https://docs.google.com/document/d/1hm_uRWeTqrN2SiqpwGJleEBLbknZ0GsEgmVxhB8eKJQ/edit?disco=AAAAJ3ADiC4&usp=comment_email_discussion&usp_dm=false&ts=5f3a92f1)

My bad didn't see that conversation! Thanks for fixing.

**Comment by wvega** [comment]
No worries. I noticed that we needed a change towards the end of the day on Friday, so that was a late addition. I just thought it made sense to fix it here instead of adding another PR.

**Pull request merged**